### ISSUE TITLE
Focus Manager - Concurrent Modification Exception

### ISSUE THREAD

**Issue by bawelter** [jifalops/datetime_picker_formfield#51]
When showing the picker there is a concurrent modification exception thrown every time.  My best guess is the Text field and datepicker are trying to get the focus when the user clicks on the text field.

datetime_picker_formfield: ^0.4.0

Flutter (Channel beta, v1.7.8+hotfix.4, on Mac OS X 10.14.5 18F132, locale en-US)
    • Flutter version 1.7.8+hotfix.4 at /Users/saltechsystems/flutter
    • Framework revision 20e59316b8 (4 days ago), 2019-07-18 20:04:33 -0700
    • Engine revision fee001c93f
    • Dart version 2.4.0

```
E/flutter ( 9546): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: Concurrent modification during iteration: Instance of '_CompactLinkedHashSet<FocusNode>'.
E/flutter ( 9546): #0      _CompactIterator.moveNext (dart:collection-patch/compact_hash.dart:443:7)
E/flutter ( 9546): #1      FocusManager._applyFocusChange (package:flutter/src/widgets/focus_manager.dart:1056:28)
E/flutter ( 9546): #2      _rootRun (dart:async/zone.dart:1120:38)
E/flutter ( 9546): #3      _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 9546): #4      _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 9546): #5      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 9546): #6      _rootRun (dart:async/zone.dart:1124:13)
E/flutter ( 9546): #7      _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 9546): #8      _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 9546): #9      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 9546): #10     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter ( 9546): #11     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
```

**Comment by jifalops** [comment]
This is the result of using [`FocusScope.of(context).requestFocus(FocusNode())`](https://github.com/jifalops/datetime_picker_formfield/blob/917358fd19130bd9f424736e0b89bf0123a7f2eb/lib/datetime_picker_formfield.dart#L273) to hide the keyboard when the `TextField` is tapped and the picker is going to be displayed. 

There may be a better approach here https://github.com/flutter/flutter/issues/7247#issuecomment-512886311

**Comment by bawelter** [comment]
You could try this instead using WidgetsBinding.instance.addPostFrameCallback and do the unfocus in the clear function / _handleFocusChanged functions.

```
  void requestUpdate() {
    if (!isShowingDialog) {
      isShowingDialog = true;

      widget.onShowPicker(context, value).then((newValue) {
        isShowingDialog = false;
        if (newValue != null) {
          _effectiveController.text = format(newValue);
        }
      });
    }
  }

  void _handleFocusChanged() {
    if (hasFocus && !hadFocus && (!hasText || widget.readOnly)) {
      _effectiveFocusNode.unfocus();
      requestUpdate();
    } else if (hadFocus && !hasFocus) {}
    hadFocus = hasFocus;
  }

  void clear() async {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _effectiveFocusNode.unfocus();
      _effectiveController.clear();
      requestUpdate();
    });
  }
```

**Comment by bawelter** [comment]
I created a pull request for the proposed solution.

**Comment by jifalops** [comment]
Thanks I'll check it out

**Comment by jifalops** [comment]
Fixed in 0.4.1 / 1.0.0-pre.2
